File: android/src/main/kotlin/cat/durban/sergio/flutter_meta_appads_sdk/FlutterMetaAppadsSdkPlugin.kt
package cat.durban.sergio.flutter_meta_appads_sdk

import cat.durban.sergio.flutter_meta_appads_sdk.helper.CurrencyFinder
import cat.durban.sergio.flutter_meta_appads_sdk.helper.FBUserData
import android.content.Context
import android.os.Bundle
import android.util.Log

import com.facebook.AccessToken
import com.facebook.FacebookSdk
import com.facebook.LoggingBehavior
import com.facebook.Profile
import com.facebook.appevents.AppEventsConstants
import com.facebook.appevents.AppEventsLogger

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import java.util.Currency
import java.util.Optional
import kotlin.jvm.optionals.getOrNull

/** FlutterMetaAppadsSdkPlugin */
class FlutterMetaAppadsSdkPlugin: FlutterPlugin, FlutterMetaAppadsSdkHostApi {
  private lateinit var eventsLogger: AppEventsLogger
  private var userData: FBUserData = FBUserData
  private lateinit var context: Context
  private var loggingEnabled: Boolean = false

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    context = flutterPluginBinding.applicationContext
    eventsLogger = AppEventsLogger.newLogger(flutterPluginBinding.applicationContext)
    
    FlutterMetaAppadsSdkHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
  }

  // MARK: - FlutterMetaAppadsSdkHostApi Implementation
  
  override fun initSdk(enableLogging: Boolean) {
    loggingEnabled = enableLogging
    
    if (loggingEnabled) {
      FacebookSdk.setIsDebugEnabled(true)
      FacebookSdk.addLoggingBehavior(LoggingBehavior.APP_EVENTS)
      FacebookSdk.addLoggingBehavior(LoggingBehavior.DEVELOPER_ERRORS)
    }

    FacebookSdk.setAutoInitEnabled(true)
    FacebookSdk.fullyInitialize()

    if (loggingEnabled) {
      Log.d("FBSDKLog", "Init SDK: ${FacebookSdk.isInitialized()}")
      Log.d("FBSDKLog", "SDK Version: ${FacebookSdk.getSdkVersion()}")
      Log.d("FBSDKLog", "AnonymousID: ${AppEventsLogger.getAnonymousAppDeviceGUID(context)}")
      Log.d("FBSDKLog", "TRACKING ENABLED: ${FacebookSdk.getAdvertiserIDCollectionEnabled()}")
      Log.d("FBSDKLog", "AUTOMATIC EVENT COLLECTION ENABLED: ${FacebookSdk.getAutoLogAppEventsEnabled()}")
    }
  }

  override fun setUserData(request: FBSetUserDataRequest) {
    setUserDataInternal(request)
  }
  
  override fun logStandardEvent(request: FBLogStandardEventRequest) {
    logStandardEventInternal(request)
  }
  
  override fun logPurchase(request: FBLogPurchaseRequest) {
    logPurchaseInternal(request)
  }
  
  override fun logEvents(request: FBLogEventRequest) {
    logEventInternal(request)
  }
  
  override fun getFbAnonId(): FBAnonIdResponse? {
    val anonId = AppEventsLogger.getAnonymousAppDeviceGUID(context).toString()
    return FBAnonIdResponse(fbAnonId = anonId)
  }
  
  override fun setAdvertiserTrackingEnabled(isEnabled: Boolean) {
    // NOTHING TO DO - EXCLUSIVE iOS FUNCTIONALITY
  }
  
  override fun setAdvertiserIDCollectionEnabled(isEnabled: Boolean) {
    FacebookSdk.setAdvertiserIDCollectionEnabled(isEnabled)

    if (loggingEnabled) {
      Log.d("FBSDKLog", "TRACKING ENABLED: ${FacebookSdk.getAdvertiserIDCollectionEnabled()}")
    }
  }
  
  override fun setAutoLogAppEventsEnabled(isEnabled: Boolean) {
    FacebookSdk.setAutoLogAppEventsEnabled(isEnabled)

    if (loggingEnabled) {
      Log.d("FBSDKLog", "AUTOMATIC EVENT COLLECTION ENABLED: ${FacebookSdk.getAutoLogAppEventsEnabled()}")
    }
  }

  override fun setDataProcessingOptions(request: FBSetDataProcessingOptionsRequest) {
    if (request.country != null && request.state != null) {
      FacebookSdk.setDataProcessingOptions(
        request.modes.toTypedArray(),
        request.country.toInt(),
        request.state.toInt(),
      )
    } else {
      FacebookSdk.setDataProcessingOptions(
        request.modes.toTypedArray(),
      )
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
  }

  private fun logEventInternal(request: FBLogEventRequest) {
    val params = Bundle()

    for (key in request.eventParameters.keys) {
      val value = request.eventParameters[key]

      params.putString(key, value)
    }

    eventsLogger.logEvent(request.eventName, params)
  }

  private fun logPurchaseInternal(request: FBLogPurchaseRequest) {
    val params = Bundle()
    
    for (key in request.eventParameter.keys) {
      val value = request.eventParameter[key]
      params.putString(key, value)
    }
    
    val currency = CurrencyFinder.find(request.currency).orElse(null)
    
    if (currency != null) {
      eventsLogger.logPurchase(request.amount.toBigDecimal(), currency, params)
    } else {
      Log.e("FlutterMetaAppAdsSdk", "logPurchaseInternal: Currency inválida: ${request.currency}")
    }
  }

  private fun logStandardEventInternal(request: FBLogStandardEventRequest) {
    val eventName = pigeonEnumToEventName(request.eventName)
    if (eventName != null) {
      val params = Bundle()

      for (elm in request.parameters) {
        val elmKey = pigeonEnumToEventParameter(elm.parameterName)

        if (elmKey != null) {
          params.putString(elmKey, elm.value)
        }
      }

      eventsLogger.logEvent(eventName, params)
    }
  }

  private fun setUserDataInternal(request: FBSetUserDataRequest) {
      userData.populateFromPigeon(request.type, request.value)

      AppEventsLogger.setUserData(
        userData.email,
        userData.firstName,
        userData.lastName,
        userData.phone,
        userData.birthDate,
        userData.gender,
        userData.city,
        userData.state,
        userData.zipCode,
        userData.country,
        userData.externalId
      )
  }
  
  // MARK: - Convert Pigeon enums to SDK Values
  
  private fun pigeonEnumToEventName(eventName: FBStandardEvent): String? {
    return when (eventName) {
      FBStandardEvent.AD_CLICK -> AppEventsConstants.EVENT_NAME_AD_CLICK
      FBStandardEvent.AD_IMPRESSION -> AppEventsConstants.EVENT_NAME_AD_IMPRESSION
      FBStandardEvent.COMPLETED_REGISTRATION -> AppEventsConstants.EVENT_NAME_COMPLETED_REGISTRATION
      FBStandardEvent.COMPLETED_TUTORIAL -> AppEventsConstants.EVENT_NAME_COMPLETED_TUTORIAL
      FBStandardEvent.CONTACT -> AppEventsConstants.EVENT_NAME_CONTACT
      FBStandardEvent.CUSTOMIZE_PRODUCT -> AppEventsConstants.EVENT_NAME_CUSTOMIZE_PRODUCT
      FBStandardEvent.DONATE -> AppEventsConstants.EVENT_NAME_DONATE
      FBStandardEvent.FIND_LOCATION -> AppEventsConstants.EVENT_NAME_FIND_LOCATION
      FBStandardEvent.RATED -> AppEventsConstants.EVENT_NAME_RATED
      FBStandardEvent.SCHEDULE -> AppEventsConstants.EVENT_NAME_SCHEDULE
      FBStandardEvent.SEARCHED -> AppEventsConstants.EVENT_NAME_SEARCHED
      FBStandardEvent.START_TRIAL -> AppEventsConstants.EVENT_NAME_START_TRIAL
      FBStandardEvent.SUBMIT_APPLICATION -> AppEventsConstants.EVENT_NAME_SUBMIT_APPLICATION
      FBStandardEvent.SUBSCRIBE -> AppEventsConstants.EVENT_NAME_SUBSCRIBE
      FBStandardEvent.VIEWED_CONTENT -> AppEventsConstants.EVENT_NAME_VIEWED_CONTENT
      FBStandardEvent.ADDED_PAYMENT_INFO -> AppEventsConstants.EVENT_NAME_ADDED_PAYMENT_INFO
      FBStandardEvent.ADDED_TO_CART -> AppEventsConstants.EVENT_NAME_ADDED_TO_CART
      FBStandardEvent.ADDED_TO_WISHLIST -> AppEventsConstants.EVENT_NAME_ADDED_TO_WISHLIST
      FBStandardEvent.INITIATED_CHECKOUT -> AppEventsConstants.EVENT_NAME_INITIATED_CHECKOUT
      FBStandardEvent.PURCHASED -> AppEventsConstants.EVENT_NAME_PURCHASED
      FBStandardEvent.ACHIEVED_LEVEL -> AppEventsConstants.EVENT_NAME_ACHIEVED_LEVEL
      FBStandardEvent.UNLOCKED_ACHIEVEMENT -> AppEventsConstants.EVENT_NAME_UNLOCKED_ACHIEVEMENT
      FBStandardEvent.SPENT_CREDITS -> AppEventsConstants.EVENT_NAME_SPENT_CREDITS
    }
  }
  
  private fun pigeonEnumToEventParameter(eventParameter: FBStandardParameter): String? {
    return when (eventParameter) {
      FBStandardParameter.PARAMETER_NAME_CONTENT -> AppEventsConstants.EVENT_PARAM_CONTENT
      FBStandardParameter.PARAMETER_NAME_CONTENT_ID -> AppEventsConstants.EVENT_PARAM_CONTENT_ID
      FBStandardParameter.PARAMETER_NAME_CONTENT_TYPE -> AppEventsConstants.EVENT_PARAM_CONTENT_TYPE
      FBStandardParameter.PARAMETER_NAME_CURRENCY -> AppEventsConstants.EVENT_PARAM_CURRENCY
      FBStandardParameter.PARAMETER_NAME_DESCRIPTION -> AppEventsConstants.EVENT_PARAM_DESCRIPTION
      FBStandardParameter.PARAMETER_NAME_LEVEL -> AppEventsConstants.EVENT_PARAM_LEVEL
      FBStandardParameter.PARAMETER_NAME_MAX_RATING_VALUE -> AppEventsConstants.EVENT_PARAM_MAX_RATING_VALUE
      FBStandardParameter.PARAMETER_NAME_NUM_ITEMS -> AppEventsConstants.EVENT_PARAM_NUM_ITEMS
      FBStandardParameter.PARAMETER_NAME_PAYMENT_INFO_AVAILABLE -> AppEventsConstants.EVENT_PARAM_PAYMENT_INFO_AVAILABLE
      FBStandardParameter.PARAMETER_NAME_REGISTRATION_METHOD -> AppEventsConstants.EVENT_PARAM_REGISTRATION_METHOD
      FBStandardParameter.PARAMETER_NAME_SEARCH_STRING -> AppEventsConstants.EVENT_PARAM_SEARCH_STRING
      FBStandardParameter.PARAMETER_NAME_SUCCESS -> AppEventsConstants.EVENT_PARAM_SUCCESS
      FBStandardParameter.PARAMETER_NAME_AD_TYPE -> AppEventsConstants.EVENT_PARAM_AD_TYPE
      FBStandardParameter.PARAMETER_NAME_ORDER_ID -> AppEventsConstants.EVENT_PARAM_ORDER_ID
      FBStandardParameter.PARAMETER_NAME_EVENT_NAME -> "fb_event_name"
      FBStandardParameter.PARAMETER_NAME_LOG_TIME -> "fb_log_time"
    }
  }
}

File: ios/flutter_meta_appads_sdk/Sources/flutter_meta_appads_sdk/FlutterMetaAppadsSdkPlugin.swift
import FBSDKCoreKit
import Flutter
import UIKit

public class FlutterMetaAppadsSdkPlugin: NSObject, FlutterPlugin, FlutterSceneLifeCycleDelegate, FlutterMetaAppadsSdkHostApi {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FlutterMetaAppadsSdkPlugin()
        
        FlutterMetaAppadsSdkHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
        
        // Register for both AppDelegate and SceneDelegate callbacks so the
        // plugin keeps working on apps that have not yet adopted the
        // UIScene lifecycle as well as those that have.
        registrar.addApplicationDelegate(instance)
        registrar.addSceneDelegate(instance)
    }

    private var startUpApplication: UIApplication?
    private var startUpURL: URL?
    private var startUpSourceApplication: String?
    private var startUpAnnotation: Any?
    private var loggingEnabled: Bool = false
    private var sdkInitialized: Bool = false

    // MARK: - UIApplicationDelegate (legacy, pre-UIScene lifecycle)

    public func application(
        _ application: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    ) -> Bool {
        let sourceApplication = options[.sourceApplication] as? String
        let annotation = options[.annotation]

        forwardOpenURL(
            application: application,
            url: url,
            sourceApplication: sourceApplication,
            annotation: annotation
        )

        // Don't short-circuit other handlers (e.g. Flutter's deep-link router).
        return false
    }

    public func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
    ) -> Bool {
        ApplicationDelegate.shared.application(application, continue: userActivity)

        // Don't short-circuit other handlers (e.g. Flutter's deep-link router).
        return false
    }

    // MARK: - FlutterSceneLifeCycleDelegate (UIScene lifecycle)

    public func scene(
        _ scene: UIScene,
        willConnectTo session: UISceneSession,
        options connectionOptions: UIScene.ConnectionOptions?
    ) -> Bool {
        if let contexts = connectionOptions?.urlContexts {
            forwardURLContexts(contexts)
        }

        return false
    }

    public func scene(
        _ scene: UIScene,
        openURLContexts URLContexts: Set<UIOpenURLContext>
    ) -> Bool {
        forwardURLContexts(URLContexts)

        return false
    }

    public func scene(_ scene: UIScene, continue userActivity: NSUserActivity) -> Bool {
        ApplicationDelegate.shared.application(UIApplication.shared, continue: userActivity)

        // Don't short-circuit other handlers (e.g. Flutter's deep-link router).
        return false
    }

    private func captureStartUp(url: URL, sourceApplication: String?, annotation: Any?) {
        startUpApplication = UIApplication.shared
        startUpURL = url
        startUpSourceApplication = sourceApplication
        startUpAnnotation = annotation
    }

    private func forwardOpenURL(
        application: UIApplication,
        url: URL,
        sourceApplication: String?,
        annotation: Any?
    ) {
        if sdkInitialized {
            ApplicationDelegate.shared.application(
                application,
                open: url,
                sourceApplication: sourceApplication,
                annotation: annotation as Any
            )
        } else {
            // Defer to initSdk() so the SDK can record the launch URL once it
            // is initialized by Dart code.
            captureStartUp(
                url: url,
                sourceApplication: sourceApplication,
                annotation: annotation
            )
        }
    }

    private func forwardURLContexts(_ contexts: Set<UIOpenURLContext>) {
        guard !contexts.isEmpty else { return }

        let application = UIApplication.shared

        for context in contexts {
            forwardOpenURL(
                application: application,
                url: context.url,
                sourceApplication: context.options.sourceApplication,
                annotation: context.options.annotation
            )
        }
    }

    // MARK: - FlutterMetaAppadsSdkHostApi Implementation
    
    func initSdk(enableLogging: Bool) throws {
        loggingEnabled = enableLogging
        initSdk()
    }
    
    func setUserData(request: FBSetUserDataRequest) throws {
        setUserData(request.value, forType: request.type)
    }
    
    func logStandardEvent(request: FBLogStandardEventRequest) throws {
        logEvent(request.eventName, parameters: request.parameters)
    }
    
    func logPurchase(request: FBLogPurchaseRequest) throws {
        logPurchase(request.amount, currency: request.currency, parameters: request.eventParameter)
    }
    
    func logEvents(request: FBLogEventRequest) throws {
        logEvent(request.eventName, parameters: request.eventParameters)
    }
    
    func getFbAnonId() throws -> FBAnonIdResponse? {
        return FBAnonIdResponse(fbAnonId: AppEvents.shared.anonymousID)
    }
    
    func setAdvertiserTrackingEnabled(isEnabled: Bool) throws {
        Settings.shared.isAdvertiserTrackingEnabled = isEnabled
        
        if loggingEnabled {
            print("FBSDKLog: TRACKING ENABLED \(Settings.shared.isAdvertiserTrackingEnabled)")
        }
    }
    
    func setAdvertiserIDCollectionEnabled(isEnabled: Bool) throws {
        Settings.shared.isAdvertiserIDCollectionEnabled = isEnabled
        
        if loggingEnabled {
            print("FBSDKLog: IDFA COLLECTION \(Settings.shared.isAdvertiserIDCollectionEnabled)")
        }
    }
    
    func setAutoLogAppEventsEnabled(isEnabled: Bool) throws {
        Settings.shared.isAutoLogAppEventsEnabled = isEnabled
        
        if loggingEnabled {
            print("FBSDKLog: AUTOMATIC EVENT COLLECTION \(Settings.shared.isAutoLogAppEventsEnabled)")
        }
    }
    
    func setDataProcessingOptions(request: FBSetDataProcessingOptionsRequest) throws {
        if let country = request.country, let state = request.state {
            Settings.shared.setDataProcessingOptions(request.modes, country: Int32(country), state: Int32(state))
        } else {
            Settings.shared.setDataProcessingOptions(request.modes)
        }
    }

    private func initSdk() {
        if loggingEnabled {
            Settings.shared.enableLoggingBehavior(.appEvents)
            Settings.shared.enableLoggingBehavior(.informational)
        }

        if let application = startUpApplication,
           let url = startUpURL {
            ApplicationDelegate.shared.application(
                application,
                open: url,
                sourceApplication: startUpSourceApplication,
                annotation: startUpAnnotation as Any
            )
        } else {
            ApplicationDelegate.shared.initializeSDK()
        }

        sdkInitialized = true
        startUpApplication = nil
        startUpURL = nil
        startUpSourceApplication = nil
        startUpAnnotation = nil

        if loggingEnabled {
            print("FBSDKLog: SDK Version: \(Settings.shared.sdkVersion)")
            print("FBSDKLog: AnonymousID: \(AppEvents.shared.anonymousID)")
            print("FBSDKLog: IDFA COLLECTION \(Settings.shared.isAdvertiserIDCollectionEnabled)")
            print("FBSDKLog: TRACKING ENABLED \(Settings.shared.isAdvertiserTrackingEnabled)")
            print("FBSDKLog: AUTOMATIC EVENT COLLECTION \(Settings.shared.isAutoLogAppEventsEnabled)")
        }
    }

    private func logEvent(_ eventName: String, parameters: [String: String]) {
        AppEvents.shared.logEvent(
            AppEvents.Name(eventName),
            parameters: convertCustomParametersToFbParameters(parameters)
        )
    }

    private func logEvent(_ eventName: FBStandardEvent, parameters: [FBStandardEventParameter]) {
        if let appEventName = pigeonEnumToEventName(eventName) {
            AppEvents.shared.logEvent(
                appEventName,
                parameters: convertStandardParametersToFbParameters(parameters)
            )
        }
    }

    private func logPurchase(_ amount: Double, currency: String, parameters: [String: String]) {
        AppEvents.shared.logPurchase(
            amount: amount,
            currency: currency,
            parameters: convertCustomParametersToFbParameters(parameters)
        )
    }

    private func setUserData(_ data: String?, forType: FBUserDataType) {
        if let forType = pigeonEnumToFBSDKAppEventUserDataType(forType) {
            AppEvents.shared.setUserData(data, forType: forType)

            print("FBSDKLog: TRACKING ENABLED \(Settings.shared.isAdvertiserTrackingEnabled)")
        }
    }
}

extension FlutterMetaAppadsSdkPlugin {
    fileprivate func convertCustomParametersToFbParameters(_ parameters: [String: String]) -> [AppEvents.ParameterName: String] {
        var fbParameters: [AppEvents.ParameterName: String] = [:]

        for parameter in parameters {
            fbParameters[AppEvents.ParameterName(rawValue: parameter.key)] = parameter.value
        }

        return fbParameters
    }

    fileprivate func convertStandardParametersToFbParameters(_ parameters: [FBStandardEventParameter]) -> [AppEvents.ParameterName: String] {
        var fbParameters: [AppEvents.ParameterName: String] = [:]

        for parameter in parameters {
            if let parameterName = pigeonEnumToEventParameter(parameter.parameterName) {
                fbParameters[parameterName] = parameter.value
            }
        }

        return fbParameters
    }
}

// MARK: - Convert enums to SDK Values

extension FlutterMetaAppadsSdkPlugin {
    fileprivate func pigeonEnumToEventName(_ eventName: FBStandardEvent) -> AppEvents.Name? {
        switch eventName {
        case .achievedLevel:
            return AppEvents.Name.achievedLevel
        case .adClick:
            return AppEvents.Name.adClick
        case .adImpression:
            return AppEvents.Name.adImpression
        case .completedRegistration:
            return AppEvents.Name.completedRegistration
        case .completedTutorial:
            return AppEvents.Name.completedTutorial
        case .contact:
            return AppEvents.Name.contact
        case .customizeProduct:
            return AppEvents.Name.customizeProduct
        case .donate:
            return AppEvents.Name.donate
        case .findLocation:
            return AppEvents.Name.findLocation
        case .rated:
            return AppEvents.Name.rated
        case .schedule:
            return AppEvents.Name.schedule
        case .searched:
            return AppEvents.Name.searched
        case .startTrial:
            return AppEvents.Name.startTrial
        case .submitApplication:
            return AppEvents.Name.submitApplication
        case .subscribe:
            return AppEvents.Name.subscribe
        case .viewedContent:
            return AppEvents.Name.viewedContent
        case .addedPaymentInfo:
            return AppEvents.Name.addedPaymentInfo
        case .addedToCart:
            return AppEvents.Name.addedToCart
        case .addedToWishlist:
            return AppEvents.Name.addedToWishlist
        case .initiatedCheckout:
            return AppEvents.Name.initiatedCheckout
        case .purchased:
            return AppEvents.Name.purchased
        case .unlockedAchievement:
            return AppEvents.Name.unlockedAchievement
        case .spentCredits:
            return AppEvents.Name.spentCredits
        }
    }

    fileprivate func pigeonEnumToEventParameter(_ eventParameter: FBStandardParameter) -> AppEvents.ParameterName? {
        switch eventParameter {
        case .parameterNameContent:
            return .content
        case .parameterNameContentID:
            return .contentID
        case .parameterNameContentType:
            return .contentType
        case .parameterNameCurrency:
            return .currency
        case .parameterNameDescription:
            return .description
        case .parameterNameLevel:
            return .level
        case .parameterNameMaxRatingValue:
            return .maxRatingValue
        case .parameterNameNumItems:
            return .numItems
        case .parameterNamePaymentInfoAvailable:
            return .paymentInfoAvailable
        case .parameterNameRegistrationMethod:
            return .registrationMethod
        case .parameterNameSearchString:
            return .searchString
        case .parameterNameSuccess:
            return .success
        case .parameterNameAdType:
            return .adType
        case .parameterNameOrderID:
            return .orderID
        case .parameterNameEventName:
            return .eventName
        case .parameterNameLogTime:
            return .logTime
        }
    }

    fileprivate func pigeonEnumToFBSDKAppEventUserDataType(_ type: FBUserDataType) -> FBSDKAppEventUserDataType? {
        switch type {
        case .email:
            return .email
        case .firstName:
            return .firstName
        case .lastName:
            return .lastName
        case .phone:
            return .phone
        case .dateOfBirth:
            return .dateOfBirth
        case .gender:
            return .gender
        case .city:
            return .city
        case .state:
            return .state
        case .zip:
            return .zip
        case .country:
            return .country
        case .externalId:
            return .externalId
        }
    }
}
